design/httpproxy-include-route-duplicates-design.md

```diff
@@ -0,0 +1,301 @@
+
+We realized we should ensure more consistent and correct handling of duplicate routes.
+We can possibly provide a better UX if the total route is considered when validating duplicates, rather than individual segments.
+However, this does come with more complications on how to report errors.
+This design is an effort to rework things and present ideas to the community.
+
+## Goals
+- Establish consistent patterns for detecting duplicate route match configurations in a tree of HTTPProxies
+- Establish status condition reporting pattern for notifying users of errors in their configuration
+- Ensure all valid user configuration is still programmed, even if there are duplicate route errors in an HTTPProxy tree or individual HTTPProxy
+
+## Non Goals
+- Address or change the behavior of HTTPProxy routes being able to override Ingress routes
+
+## High-Level Design
+Rather than checking for duplicates at the include conditions level, we will do route match condition duplicate validation at the "leaves" of the include/route tree to consider a full route.
+In the case of duplicate routes, precedence will be given to the route that comes "first", criteria described in more detail below.
+A HTTPProxy that contains a duplicate route will have an error condition on it, be marked invalid, and warning conditions set on its parents, up to the root.
+Any error/warning conditions will contain salient details for owners of the various affected resources to use to debug/remediate the issue.
+We will attempt to program as much valid configuration as we can, i.e. don't short circuit processing an HTTPProxy because part of it is invalid.
+
+## Detailed Design
+
+### Single HTTPProxy
+
+Currently the following HTTPProxy is allowed and considered valid:
+
+```
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: example
+spec:
+  virtualhost:
+    fqdn: example.com
+  routes:
+  - conditions:
+    - prefix: /foo
+    services:
+    - name: s1
+      port: 8080
+  - conditions:
+    - prefix: /foo
+    services:
+    - name: s2
+      port: 8080
+```
+
+The actual route that is programmed will forward requests for `example.com/foo` to `s2`.
+
+Instead, this design proposes that since the route to `s1` comes first, that route will be programmed and the second that routes to `s2` will be considered a duplicate.
+Requests to `example.com/foo` will be routed to `s1`.
+The `example` HTTPProxy will be set with Status below:
+
+```
+status:
+  conditions:
+  - errors:
+    - message: duplicate match conditions defined on route matching path prefix: "/foo"
+      reason: DuplicateMatchConditions
+      status: "True"
+      type: RouteError
+    lastTransitionTime: ...
+    message: At least one error present, see Errors for details
+    observedGeneration: ...
+    reason: ErrorPresent
+    status: "False"
+    type: Valid
+  currentStatus: invalid
+  description: At least one error present, see Errors for details
+  ...
```

Comment:
nit: I cant quite remember the contour standard, but since we program the route anyway, would a warning be more apt? 
just from a UX perspective an error come across as non-recoverable and implies that none of the routes were programmed.